### ISSUE TITLE
False positive unrelated_type_equality_checks for certain generics cases

### ISSUE THREAD

**Issue by camsteffen** [dart-lang/linter#1572]
I found a false positive for unrelated_type_equality_checks similar to #1297 but more specific.

```
class Foo<T> {
  Bar<T> bar;

  bool test(Baz baz) => bar == baz; // false positive here
}

class Bar<T> {}

class Baz extends Bar<int> {}
```

**Comment by pq** [comment]
fyi @srawlins who's fixed a bunch of issues in this one recently.

**Comment by srawlins** [comment]
Hey now, I never said I _enjoyed_ working on unrelated_type_equality_checks. 😛
